components/board/genericlinux/gpio.go

```diff
@@ -277,7 +295,9 @@ func (pin *gpioPin) halfPwmCycle(shouldBeOn bool) bool {
 		dutyCycle = 1 - dutyCycle
 	}
 	duration := time.Duration(float64(time.Second) * dutyCycle / float64(freqHz))
-	return accurateSleep(pin.cancelCtx, duration)
+	pin.swPWMContext, pin.swPWMCancel = context.WithCancel(pin.cancelCtx)
```

Comment:
This is going to create new contexts every time we toggle the pin, because it's inside `halfPwmCycle()`. I think instead, we should create new ones only when we set `swPwmRunning` to true (in effect, we can replace `swPwmRunning` with `swPwmContext != nil` and get rid of `swPwmRunning` entirely). 

Even that doesn't quite seem right: if some code replaces `pin.swPwmContext` with a new one while `halfPwmCycle()` isn't looking, `halfPwmCycle()` will start checking the new one and ignoring the old one.

What we kinda want is that every time we call `softwarePwmLoop` we pass a new context as an argument to it, and it checks that one instead of doing anything with a field stored in the struct.

```go
func (pin *gpioPin) startSoftwarePWM() error {                                                      
    ...

    // If we get here, we need a software loop to drive the PWM signal.
    if pin.swPwmCancel != nil {
        return nil // One is already running.
    }
    
    swPwmContext, swPwmCancel = context.WithCancel(pin.cancelCtx)
    pin.swPwmCancel  = swPwmCancel
    pin.boardWorkers.Add(1)                                                                      
    utils.ManagedGo(func() {pin.softwarePwmLoop(swPwmContext)}, pin.boardWorkers.Done)
    return nil
}

// softwarePwmLoop and halfPwmCycle both take a new argument, the context that might be canceled elsewhere.

Reply:
implemented this but I think we will still need to store the swPwmContext in the struct because we should use the swPwmContext in `accurateSleep` instead of the pin context. lmk If you think this is incorrect?

Reply:
I think you're right that we need to use swPwmContext in accurateSleep. but I think we can pass it around as a function argument, rather than relying on mutable state within the struct. but I'm open to other possibilities on that front!